### PR TITLE
WIP Prototype of reliable delivery in Typed, #20984

### PR DESCRIPTION
**Pull request by patriknw**
I had some spare time in the weekend to create a prototype of support for reliable delivery in Akka Typed. I'm taking a rather different approach than in untyped `AtLeastOnceDelivery` because I would like:

* support plain flow control, with a work pulling approach
* be possible to use with or without persistence (without it may loose messages if producer node crashes)
* detect lost messages on the consumer side and let that drive resends, instead of aggressively resending from producer side
* deliver messages in order
* have an efficient protocol for acknowledgments over the network (not ack each message), but still have a simple one-by-one protocol for the end user

I can imagine that it might be somewhat overwhelming to directly dive into the full 400 lines so I'd recommend that you first look at the [first commit](https://github.com/akka/akka/commit/5e35c4c63b105de9092bd80af4ffc659e9551a56) to understand how the work pulling works, then add the [resending of lost messages](https://github.com/akka/akka/commit/5ff163e684fb7bfc1f6840be70679c896538793a). Those two are written in application domain code.

Then I [extracted the Akka code](https://github.com/akka/akka/commit/e506dc52622f8f5ec452a5da2f3ce3084782f85c) into ProducerController and ConsumerController.

A sequence diagram illustrating the flow control protocol:

reliabledelivery

Refs #20984

### CODE REVIEW

File: akka-actor-typed/src/main/scala/akka/actor/typed/delivery/ReliableDeliveryPoc.scala [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+    final case class SequencedMessage[T](seqNr: Long, msg: T) extends ConsumerMessage
+    private final case object RetryRequest extends ConsumerMessage
+
+    // FIXME name?
+    final case class Delivery[T](seqNr: Long, msg: T, confirmTo: ActorRef[Confirmed[T]])
+    final case class Confirmed[T](seqNr: Long, deliverNextTo: ActorRef[Delivery[T]]) extends ConsumerMessage
+
+    private val RequestWindow = 50
+
+    def behavior[T](producer: ActorRef[ProducerMessage], resendLost: Boolean): Behavior[ConsumerMessage] = {
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/delivery/ReliableDeliveryPoc.scala]: this is the behavior that we would also provide as a persistent variant

File: akka-actor-typed/src/main/scala/akka/actor/typed/delivery/ReliableDeliveryPoc.scala [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+
+    sealed trait ProducerMessage
+    final case class Request[T](confirmedSeqNr: Long, upToSeqNr: Long, consumer: ActorRef[SequencedMessage[T]],
+                                supportResend: Boolean, viaReceiveTimeout: Boolean) extends ProducerMessage {
+      require(confirmedSeqNr < upToSeqNr)
+    }
+    final case class Resend(fromSeqNr: Long) extends ProducerMessage
+    private case class Msg[T](msg: T) extends ProducerMessage
+
+    def behavior[T: ClassTag, RequestNext](
```
Comment (patriknw) [inline akka-actor-typed/src/main/scala/akka/actor/typed/delivery/ReliableDeliveryPoc.scala]: this is the behavior that we would also provide as a persistent variant

**Comment by johanandren** [comment]
I think this is pretty cool and that it is definitely worth continuing the work along these lines.

**Comment by patriknw** [comment]
I think this is a very interesting area to explore more, but closing the PR for now since I'm not actively working on it.

**Pull request closed without merging**